src/core/classes/TransactionTask.ts

```diff
@@ -0,0 +1,124 @@
+import { ContractWrapper } from "./contract-wrapper";
+import { BigNumber, CallOverrides, ContractTransaction } from "ethers";
+import { ContractEncoder } from "./contract-encoder";
+import { GasCostEstimator } from "./gas-cost-estimator";
+import { TransactionResult } from "../types";
+
+class TransactionTaskBuilder {
+  public contractWrapper: ContractWrapper<any>;
+  public functionName: string;
+  public args: any[] = [];
+  public overrides: CallOverrides | undefined;
+
+  constructor(contractWrapper: ContractWrapper<any>, functionName: string) {
+    this.contractWrapper = contractWrapper;
+    this.functionName = functionName;
+  }
+
+  public withArgs(...args: any[]): TransactionTaskBuilder {
+    this.args = args;
+    return this;
+  }
+
+  public withOverrides(overrides: CallOverrides): TransactionTaskBuilder {
+    this.overrides = overrides;
+    return this;
+  }
+
+  public build(): TransactionTask {
+    return new TransactionTask(this);
+  }
+
+  // TODO gasless
+}
+
+export class TransactionTask {
+  static builder(contractWrapper: ContractWrapper<any>, functionName: string) {
+    return new TransactionTaskBuilder(contractWrapper, functionName);
+  }
+
+  private contractWrapper: ContractWrapper<any>;
+  private functionName: string;
+  private args: any[] = [];
+  private overrides: CallOverrides | undefined;
+  private encoder: ContractEncoder<any>;
+  private estimator: GasCostEstimator<any>;
+
+  constructor(builder: TransactionTaskBuilder) {
+    this.contractWrapper = builder.contractWrapper;
+    this.functionName = builder.functionName;
+    this.args = builder.args;
+    this.overrides = builder.overrides;
+    this.encoder = new ContractEncoder(this.contractWrapper);
+    this.estimator = new GasCostEstimator(this.contractWrapper);
+  }
+
+  // ////////////// Overrides ////////////////
+
+  public overrideGasLimit(gasLimit: BigNumber): TransactionTask {
+    this.overrides = {
+      ...this.overrides,
+      gasLimit,
+    };
+    return this;
+  }
+
+  public overrideGasPrice(gasPrice: BigNumber): TransactionTask {
+    this.overrides = {
+      ...this.overrides,
+      gasPrice,
+    };
+    return this;
+  }
+
+  public overrideNonce(nonce: BigNumber): TransactionTask {
+    this.overrides = {
+      ...this.overrides,
+      nonce,
+    };
+    return this;
+  }
+
+  public overrideValue(value: BigNumber): TransactionTask {
+    this.overrides = {
+      ...this.overrides,
+      value,
+    };
+    return this;
+  }
+
+  // ////////////// Estimates ////////////////
+
+  public async estimateGasLimit(): Promise<BigNumber> {
+    return await this.estimator.gasLimitOf(this.functionName, this.args);
+  }
+
+  public async estimateGasCostInEther(): Promise<string> {
+    return await this.estimator.gasCostOf(this.functionName, this.args);
+  }
+
+  // ////////////// Actions ////////////////
+
+  public async encode(): Promise<string> {
```

Comment:
Can we get methods off this to get all parts of tx data - this `encode` function gives you the `data` param, but would be nice to get everything including `nonce`, `value`, `gasLimit`, etc. all in one structure from a function (similar to how fractal needs it). This would actually let us create this generic contract encoder interface for every function call.

Reply:
Here's an example of what the structure looks like: https://buildthirdweb.slack.com/archives/C03GRV58Y13/p1658382917
<img width="478" alt="Screen Shot 2022-08-05 at 12 45 04 PM" src="https://user-images.githubusercontent.com/64697628/183149745-e9db369d-08be-4a23-b88a-4b16d693f1eb.png">

And implementation in Go: https://buildthirdweb.slack.com/archives/C03GRV58Y13/p165844326355217
<img width="209" alt="Screen Shot 2022-08-05 at 12 45 33 PM" src="https://user-images.githubusercontent.com/64697628/183149811-96016659-0917-4759-b3d9-48e041082d39.png">

Reply:
But maybe most convenient is just to have a single function that returns the entire tx object, however ethers expects it, so people can do signing themselves as well (or at least for now to access the data).

Reply:
yeah that's interesting basically make this one `encodeFunctionData()` and add another `encodeTransaction()` or something like that